pymodbus/transport/transport.py

```diff
@@ -605,6 +613,8 @@ def sendto(self, data: bytes, _addr: Any = None) -> None:
 
     def write(self, data: bytes) -> None:
         """Send data"""
+        if not self.other_modem:
```

Comment:
In that case, I'll need to use `type: ignore` instead of runtime checks.

Reply:
??? self.other_modem is a reference to the other part. The object is created as part of the connect process.

Reply:
I understand what `self.other_modem` is.  However, there is currently no hard guarantee that this function is not called with `self.other_modem == None`.

Options:
(1) add this check
(2) `type: ignore` and trust all calling code to be correct (such a "soft" guarantee will not please mypy)
(3) refactor the object/methods so `write()` can only be called after connect (?!)

Reply:
The hard guarantee is that it is our own code !

(1) is not really the way we want to go !
(2) this is why I do not like mypy !
(3) but that is already the fact !, that class is instantiated by the connect call....the problem is that mypy sees this as an independent class which it is NOT.

Mypy needs to learn (and I have absolutely no idea how) that some class can only be instantiated in a context, and not randomly. The solution might be to secure that other_modem can never be null (like you suggested for transport).

I do believe that mypy helps a lot in making the code more stable, but I refuse to make "bad" code just because mypy does not understand e.g. that some classes cannot be used randomly.

To put it in a positive light, I approve any type change you make without a second thought, including small non essential code changes !! but when it affects the running code like asserting due to a programming error on our side, I am strongly opposed. 

I think you do a very good and important job, so please do not see my comments as solely negative !

Reply:
https://github.com/pymodbus-dev/pymodbus/pull/1873